hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSyncUp.java

```diff
@@ -69,41 +80,103 @@ public static void main(String[] args) throws Exception {
+      ReplicationQueueId queueId = iter.next();
+      if (!queueId.isRecovered()) {
+        existingQueuePeerIds.add(queueId.getPeerId());
+      }
+    }
+
+    for (String peerId : peerIds) {
+      if (!existingQueuePeerIds.contains(peerId)) {
+        ReplicationQueueId queueId = new ReplicationQueueId(regionServer, peerId);
+        System.out.println("Add replication queue " + queueId + " for claiming");
+        storage.setOffset(queueId, regionServer.toString(), ReplicationGroupOffset.BEGIN,
+          Collections.emptyMap());
+      }
+    }
+  }
+
+  private void addMissingReplicationQueues(ReplicationQueueStorage storage,
+    Set<ServerName> regionServers, Set<String> peerIds) throws ReplicationException {
+    for (ServerName regionServer : regionServers) {
+      addMissingReplicationQueues(storage, regionServer, peerIds);
+    }
   }
 
   // When using this tool, usually the source cluster is unhealthy, so we should try to claim the
   // replication queues for the dead region servers first and then replicate the data out.
-  private void claimReplicationQueues(ZKWatcher zkw, ReplicationSourceManager mgr)
-    throws ReplicationException, KeeperException {
-    // TODO: reimplement this tool
-    // List<ServerName> replicators = mgr.getQueueStorage().getListOfReplicators();
-    // Set<ServerName> liveRegionServers = getLiveRegionServers(zkw);
-    // for (ServerName sn : replicators) {
-    // if (!liveRegionServers.contains(sn)) {
-    // List<String> replicationQueues = mgr.getQueueStorage().getAllQueues(sn);
-    // System.out.println(sn + " is dead, claim its replication queues: " + replicationQueues);
-    // for (String queue : replicationQueues) {
-    // mgr.claimQueue(sn, queue);
-    // }
-    // }
-    // }
+  private void claimReplicationQueues(ReplicationSourceManager mgr, Set<ServerName> regionServers)
+    throws ReplicationException, KeeperException, IOException {
+    // union the region servers from both places, i.e, from the wal directory, and the records in
+    // replication queue storage.
+    Set<ServerName> replicators = new HashSet<>(regionServers);
+    ReplicationQueueStorage queueStorage = mgr.getQueueStorage();
+    replicators.addAll(queueStorage.listAllReplicators());
+    FileSystem fs = CommonFSUtils.getCurrentFileSystem(getConf());
+    Path infoDir = new Path(CommonFSUtils.getRootDir(getConf()), INFO_DIR);
+    for (ServerName sn : replicators) {
+      List<ReplicationQueueId> replicationQueues = queueStorage.listAllQueueIds(sn);
+      System.out.println(sn + " is dead, claim its replication queues: " + replicationQueues);
+      // record the rs name, so when master restarting, we will skip claiming its replication queue
+      fs.createNewFile(new Path(infoDir, sn.getServerName()));
+      for (ReplicationQueueId queueId : replicationQueues) {
+        mgr.claimQueue(queueId, true);
+      }
+    }
   }
 
   @Override
   public int run(String[] args) throws Exception {
```

Comment:
Whether there is a problem with starting two ReplicationSyncUp processes at the same time?
There is no check in the code at present.

Reply:
This is a nice to have, as we do not support this in the old tool either. Can be a separate issue?